File: mobile/src/bindings/daypart.ts
import { invoke } from "@tauri-apps/api/core";
import type { DaypartView } from "./types";

export const daypart = {
  async add(name: string): Promise<void> {
    await invoke("daypart", { action: "Add", name });
  },

  async remove(name: string): Promise<void> {
    await invoke("daypart", { action: "Remove", name });
  },

  async view(name: string): Promise<DaypartView> {
    return await invoke("daypart", { action: "View", name });
  },

  async list(): Promise<string[]> {
    return await invoke("daypart", { action: "List" });
  },

  async assign(mealName: string, dayparts: string[]): Promise<void> {
    await invoke("daypart", {
      action: "Assign",
      mealname: mealName,
      dayparts,
    });
  },

  async unassign(mealName: string, dayparts: string[]): Promise<void> {
    await invoke("daypart", {
      action: "Unassign",
      mealname: mealName,
      dayparts,
    });
  },
};

File: mobile/src-tauri/src/commands/daypart.rs
use foody_core::app::App;
use serde::Deserialize;
use tauri::State;

use crate::commands::serialize_json;

#[derive(Deserialize)]
#[serde(tag = "action")]
pub enum DaypartCommand {
    Add { name: String },
    Remove { name: String },
    View { name: String },
    List,
    Assign { mealname: String, dayparts: Vec<String> },
    Unassign { mealname: String, dayparts: Vec<String> },
}

#[tauri::command]
pub async fn daypart(
    state: State<'_, App>,
    command: DaypartCommand,
) -> Result<serde_json::Value, String> {
    match command {
        DaypartCommand::Add { name } => serialize_json(state.add_daypart(&name)).await,
        DaypartCommand::Remove { name } => serialize_json(state.remove_daypart(&name)).await,
        DaypartCommand::View { name } => serialize_json(state.view_daypart(&name)).await,
        DaypartCommand::List => serialize_json(state.list_dayparts()).await,
        DaypartCommand::Assign { mealname, dayparts } => {
            serialize_json(state.assign_dayparts(&mealname, &dayparts)).await
        }
        DaypartCommand::Unassign { mealname, dayparts } => {
            serialize_json(state.unassign_dayparts(&mealname, &dayparts)).await
        }
    }
}
